### PR TITLE
refactor: document coinbase rules, refactor to ease understanding

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

Recently, zcashd had a [bug](https://github.com/zcash/zcash/pull/5804) related to coinbase transactions.

Also recently, the spec has been update to explicitly include or clarify some consensus rules related to coinbase transactions.

For both of those reasons I decided to review our implementation of those rules. 

### Specifications

The rules are spread out in the spec, but have been condensed in the change history:

image

https://zips.z.cash/protocol/protocol.pdf#2022.3.0

## Solution

- Double-checked the zcashd bug. We validate coinbase rules in totally separated manner so we are not affected by it.
- Documented the coinbase consensus rules
- Renamed some methods related to coinbase transactions to hopefully make it clearer.

What helped me wrap my head around the rules is realizing that there are 3 possible cases:

- The transaction is a coinbase transaction (has a single transparent input with null prevout)
- The transaction is a non-coinbase transaction (all transparent inputs have non-null prevouts)
- Invalid (some transparent inputs have null prevouts, and it's not a coinbase transaction)

The new names try to reflect that classification. In particular, `is_coinbase` is very well-defined in the spec so it makes sense to use it.

## Review

Anyone can review this, but if I recall correctly, @teor2345 suggesting changing the name of the `is_coinbase` method, which I changed back in this PR for the above reasons, so they might be interested. Feedback is welcome on the name changes, I'm happy to revert those and leave only the documentation if desired.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebra-chain/src/transaction.rs [hunk 1]
```diff
@@ -476,20 +476,26 @@ impl Transaction {
...
+    /// transaction, that is, does not have any coinbase input (non-null prevouts).
+    ///
+    /// Note that it's possible for a transaction return false in both
+    /// [`Transaction::is_coinbase`] and [`Transaction::is_valid_non_coinbase`],
+    /// though those transactions will be rejected.
+    pub fn is_valid_non_coinbase(&self) -> bool {
         self.inputs()
             .iter()
-            .any(|input| matches!(input, transparent::Input::Coinbase { .. }))
+            .all(|input| matches!(input, transparent::Input::PrevOut { .. }))
```
Comment (conradoplg) [inline zebra-chain/src/transaction.rs]: This is basically the negation of the old method.

File: zebra-consensus/src/block.rs [hunk 2]
```diff
@@ -180,17 +180,13 @@ where
             let now = Utc::now();
             check::time_is_valid_at(&block.header, now, &height, &hash)
                 .map_err(VerifyBlockError::Time)?;
-            check::coinbase_is_first(&block)?;
-            let coinbase_tx = block
-                .transactions
-                .get(0)
-                .expect("must have coinbase transaction");
+            let coinbase_tx = check::coinbase_is_first(&block)?;
```
Comment (conradoplg) [inline zebra-consensus/src/block.rs:183]: This change is not directly related but it seemed a shame not to do it.

File: zebra-consensus/src/transaction.rs [hunk 3]
```diff
@@ -313,15 +313,17 @@ where
...
 
-            if req.is_mempool() && tx.has_any_coinbase_inputs() {
+            if req.is_mempool() && tx.is_coinbase() {
                 return Err(TransactionError::CoinbaseInMempool);
             }
-            if tx.has_valid_coinbase_transaction_inputs() {
+            if tx.is_coinbase() {
                 check::coinbase_tx_no_prevout_joinsplit_spend(&tx)?;
+            } else if !tx.is_valid_non_coinbase() {
+                return Err(TransactionError::NonCoinbaseHasCoinbaseInput);
```
Comment (conradoplg) [inline zebra-consensus/src/transaction.rs:322]: The previous logic had exactly the same effect, but it seemed harder to understand.

File: zebra-consensus/src/transaction/check.rs [hunk 4]
```diff
@@ -124,10 +125,8 @@ pub fn has_enough_orchard_flags(tx: &Transaction) -> Result<(), TransactionError
 ///
 /// <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>
 pub fn coinbase_tx_no_prevout_joinsplit_spend(tx: &Transaction) -> Result<(), TransactionError> {
-    if tx.has_valid_coinbase_transaction_inputs() {
-        if tx.contains_prevout_input() {
-            return Err(TransactionError::CoinbaseHasPrevOutInput);
```
Comment (conradoplg) [inline zebra-consensus/src/transaction/check.rs:129]: As pointed out in the Change History in the spec, this check is redundant per the definition of "coinbase transaction".

File: zebra-consensus/src/block/check.rs [hunk 5]
```diff
@@ -15,28 +15,46 @@ use crate::{error::*, parameters::SLOW_START_INTERVAL};
...
+/// and if that coinbase transaction is the first transaction in the block.
+/// Returns the coinbase transaction is successful.
 ///
 /// "The first (and only the first) transaction in a block is a coinbase
 /// transaction, which collects and spends any miner subsidy and transaction
 /// fees paid by transactions included in this block." [§3.10][3.10]
 ///
 /// [3.10]: https://zips.z.cash/protocol/protocol.pdf#coinbasetransactions
-pub fn coinbase_is_first(block: &Block) -> Result<(), BlockError> {
+pub fn coinbase_is_first(block: &Block) -> Result<Arc<transaction::Transaction>, BlockError> {
```
Comment (teor2345) [inline zebra-consensus/src/block/check.rs:27]: Optional: this name might be easier to understand, or we could use `get_coinbase`?

```suggestion
pub fn get_valid_coinbase(block: &Block) -> Result<Arc<transaction::Transaction>, BlockError> {
```

File: zebra-consensus/src/block/check.rs [hunk 6]
```diff
@@ -15,28 +15,46 @@ use crate::{error::*, parameters::SLOW_START_INTERVAL};
...
     let mut rest = block.transactions.iter().skip(1);
-    if !first.has_valid_coinbase_transaction_inputs() {
+    if !first.is_coinbase() {
         return Err(TransactionError::CoinbasePosition)?;
     }
-    if rest.any(|tx| tx.has_any_coinbase_inputs()) {
+    // > A transparent input in a non-coinbase transaction MUST NOT have a null prevout
+    //
+    // <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>
+    if rest.any(|tx| !tx.is_valid_non_coinbase()) {
```
Comment (teor2345) [inline zebra-consensus/src/block/check.rs]: Question: Is this refactor easier to understand? I'm not sure.

```suggestion
    if !rest.all(|tx| tx.is_valid_non_coinbase()) {
```

File: zebra-consensus/src/transaction/check.rs (hunk 4 above)
Comment (teor2345) [reply zebra-consensus/src/transaction/check.rs:129]: Question: Do we want to `assert!(!tx.contains_prevout_input());` here?

**Review by teor2345 — COMMENTED** [review]
Looks good!

I have a few optional suggestions, and I'd like someone else to double-check these changes.

File: zebra-consensus/src/block/check.rs (hunk 6 above)
Comment (conradoplg) [reply zebra-consensus/src/block/check.rs]: I like it, it sounds more similar to the rule

File: zebra-consensus/src/block/check.rs (hunk 5 above)
Comment (conradoplg) [reply zebra-consensus/src/block/check.rs:27]: I'm not sure, the main purpose is checking if the coinbase is the first tx, getting it is a side effect

File: zebra-consensus/src/transaction/check.rs (hunk 4 above)
Comment (conradoplg) [reply zebra-consensus/src/transaction/check.rs:129]: I think it doesn't add much value since the `is_coinbase()` check is right before and it ensures that this is true.

I also realized that `contains_prevout_input()` is no longer being used, so that's another small reason to not do the assert. (I went ahead and removed the method it in 0ec750fdb)

**Review by oxarbitrage — DISMISSED** [review]
Looks good to me.

**Comment by teor2345** [comment]
Failed with:
> Timeout, server 34.136.108.250 not responding.
> 
> Recommendation: To check for possible causes of SSH connectivity issues and get
> recommendations, rerun the ssh command with the --troubleshoot option. Example:
> 
> gcloud None compute ssh example-instance --zone=us-central1-a --troubleshoot
> 
> Or, to investigate an IAP tunneling issue:
> 
> gcloud None compute ssh example-instance --zone=us-central1-a --tunnel-through-iap --troubleshoot
> 
> ERROR: (gcloud.compute.ssh) [/usr/bin/ssh] exited with return code [255].
https://github.com/ZcashFoundation/zebra/runs/5874622842?check_suite_focus=true

This is probably a temporary network error.

But it's weird that the PR is ready for merge anyway, we should fix that.

**Comment by teor2345** [comment]
> But it's weird that the PR is ready for merge anyway, we should fix that.

Aha, that's fixed by #3968, which is in the merge queue, and should be queued with this PR.

File: zebra-consensus/src/transaction/check.rs (hunk 4 above)
Comment (teor2345) [reply zebra-consensus/src/transaction/check.rs:129]: `TransactionError::CoinbaseHasPrevOutInput` is now unused, can you remove it as well?

Comment (conradoplg) [reply zebra-consensus/src/transaction/check.rs:129]: Done in b6ce777e2

File: zebra-consensus/src/block/check.rs [hunk 7]
```diff
@@ -15,28 +15,46 @@ use crate::{error::*, parameters::SLOW_START_INTERVAL};
...
-/// Returns `Ok(())` if there is exactly one coinbase transaction in `Block`,
-/// and that coinbase transaction is the first transaction in the block.
+/// Checks if there is exactly one coinbase transaction in `Block`,
+/// and if that coinbase transaction is the first transaction in the block.
+/// Returns the coinbase transaction is successful.
 ///
 /// "The first (and only the first) transaction in a block is a coinbase
 /// transaction, which collects and spends any miner subsidy and transaction
 /// fees paid by transactions included in this block." [§3.10][3.10]
 ///
```
Comment (daira) [inline zebra-consensus/src/block/check.rs:25]: It's section 3.11 now, and this is what it says:
> A transaction that has a single transparent input with a null prevout field, is called a coinbase transaction. Every block has a single coinbase transaction as the first transaction in the block. [...]

**Review by daira — DISMISSED** [review]
Looks good (I only checked the changes).

File: zebra-consensus/src/block/check.rs (hunk 7 above)
Comment (conradoplg) [reply zebra-consensus/src/block/check.rs:25]: Thanks for catching this, fixed in 3659a7c68

**Review by teor2345 — APPROVED** [review]
Looks good!

**Comment by teor2345** [comment]
@mergifyio refresh
